### PR TITLE
Only compare pydantic fields in `BaseModel.__eq__` instead of whole `__dict__`

### LINKED ISSUE #7444: Pydantic `__eq__` and `__hash__` doesn't work properly with `functools.cached_property`
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Bug

While porting an app to Pydantic v2, I notice that the implementation of `__eq__` doesn't work properly on Model with `functools.cached_property`: equal instances can compare unequal.

From the documentation on [computed fields](https://docs.pydantic.dev/latest/usage/computed_fields/), it seems pydantic should support `functools.cached_property`. I'm trying to compute an attribute value lazily and only once, because it is very expensive. `functools.cached_property` seems like the perfect tool for that.

Example 1 below demonstrate the problem, and Example 2 shows that this can introduce subtle bugs when using dict on "equal" objects, as dict relies on a correct `__eq__` and `__hash__`.

### Analysis

Line 831 from pydantic's implementation of `__eq__` seems to assumes that `__dict__` only ever contains pydantic's field values. 
https://github.com/pydantic/pydantic/blob/c6e2f89f36805aef03194c366afb30faa798d188/pydantic/main.py#L829-L834

I think this is not true in the general case, as `__dict__` is an instance state that is shared between pydantic and all other mechanism that need to act on the instance itself.
Indeed, `functools.cached_property` stores the computed value in the instance `__dict__`, because it is the only place it can reasonably store it:

https://github.com/python/cpython/blob/1f885df2a580360c5de69cc41191f3c6bfaaeb35/Lib/functools.py#L1005

Assuming that `__dict__` only contains pydantic's own fields has already been mentionned in #6422

### Possible fix

I think the fix below limits the comparison of `__dict__` to pydantic's own fields. I don't know much about pydantic source code, so there might be some rough edges I overlooked. I used a sentinel object as a default value to avoid crashing on missing fields, not sure it is actually necessary.

```python

    def __eq__(self, other):
        ...
        sentinel = object()
        return (
            self_type == other_type
            and all(
                getattr(self, field, sentinel) == getattr(other, field, sentinel)
                for field in self.model_fields
            )
            and self.__pydantic_private__ == other.__pydantic_private__
            and self.__pydantic_extra__ == other.__pydantic_extra__
        )
```

### Related issues
- #6422 had a more obscure problem due to assuming `__dict__` contains only pydantic fields

### Additional context
Comparing the full `__dict__` seems to be [a change from pydantic V1](https://docs.pydantic.dev/latest/migration/#changes-to-pydanticbasemodel). I would be interested in knowing why the full `__dict__` is compared in V2.

Both `attrs` and the stdlib `dataclasses` work seamlessly with `functools.cahced_property` and only compare their own fields:
#### dataclasses `__eq__`
https://github.com/python/cpython/blob/1f885df2a580360c5de69cc41191f3c6bfaaeb35/Lib/dataclasses.py#L1086-L1099

#### attrs `__eq__`
https://www.attrs.org/en/stable/comparison.html#comparison

### Example Code

```Python
# ---------- Example 1 -------------
import functools

import pydantic
    

class Model(pydantic.BaseModel):
    attr: int

    @functools.cached_property
    def cached(self) -> int:
        return 0


obj1 = Model(attr=1)
obj2 = Model(attr=1)

assert obj1 == obj2  # works

obj1.cached

assert obj1 == obj2  # raises an error

# ---------- Example 2 -------------

import functools

import pydantic


class Model(pydantic.BaseModel):
    model_config = pydantic.ConfigDict(frozen=True)

    attr: int

    @functools.cached_property
    def cached(self) -> int:
        return 0




obj1 = Model(attr=1)
obj2 = Model(attr=1)
d = {obj1: True}

assert d[obj2]

obj1.cached

assert d[obj2]  # raise an error
```

### Python, Pydantic & OS Version

```Text
Note: installed pydantic with "pip install git+https://github.com/pydantic/pydantic.git"


             pydantic version: 2.3.0
        pydantic-core version: 2.7.0
          pydantic-core build: profile=release pgo=false
                 install path: ~/miniconda3/envs/dev-sandbox/lib/python3.8/site-packages/pydantic
               python version: 3.8.17 | packaged by conda-forge | (default, Jun 16 2023, 07:11:34)  [Clang 14.0.6 ]
                     platform: macOS-10.16-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
```

### PR DESCRIPTION
**Pull request by QuentinSoubeyranAqemia**
## Change Summary

This PR fixes #7444. It is focused on fixing `pydantic.BaseModel.__eq__`, see #7786 for `pydantic.BaseModel.__hash__`.

## Related issue number

- Fix #7444 

## Checklist

* [X] The pull request title is a good summary of the changes - it will be used in the changelog
* [X] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @lig

### CODE REVIEW

**Comment by QuentinSoubeyranAqemia** [comment]
I performed a benchmark on various implementations, here are the results:

eq-benchmark_python-3-8-18-final-0

- All performance report as expressed as a ratio of the current performance, dubbed "baseline"
- I had to comment out the dict-comprehension implementation, it was too slow to run the benchmark in a timely manner
- Having a fast-path allows to not degrade performance in the common case (left column, fast-path line is close to baseline)
- To trigger the specialized code, the test fakes a `functools.cached_property` by manually editting the instances' `__dict__`. Of note, I had to add the same key to both instances, with different values, else the default `__dict__` comparison is very fast due to the dict not having the same key set. This comparison is unfair, because the bugged default implem has a fast-path. The different values are necessary to trigger the specialized code.
- If the fast-path allows to have unchanged performance in the default case, it appears to have performance cost in the functools case. This is IMHO an artifact of the testing: in the real case:
   - **either** the two object didn't pre-compute the same set of cached properties. Then, the keys of the left and right instances' `__dict__` are different, and thus the `__dict__` comparison is very fast (see point above) and doesn't cost a full dict-comparison
   - **or** the two object computed the same set of cached properties. Then, if there fields are equals, their cached properties are usually equals (cached properties are a function of the actual fields), and thus the fast-path doesn't trigger.

All that being said, I think @alexmojaki 's ideas of a fast-path and `operator.itemgetter` constitue a good performance compromise.

**Comment by QuentinSoubeyranAqemia** [comment]
The fix cause some edge-case to fail. Using the deprecated `pydantic.BaseModel.copy()` method can yield a weird instance where some fields with a default value are missing.

I'm unclear on what to do here. I could use `collections.ChainMap` to chain the instance's `__dict__` and the class' `__dict__`, but perhaps that is a problem in `pydantic.BaseModel.copy` ?
The test seems to be making sure some deprecated behavior still works, so I'd be happy to have some inputs from pydantic devs as to what the best course of action is.

**Comment by alexmojaki** [comment]
It's not just about deprecated behaviour. Pydantic allows deleting attributes even if they don't have a default:

```python
from pydantic import BaseModel

class A(BaseModel):
    b: int

a = A(b=1)
print(a.__dict__)
del a.b
print(a.__dict__)
```

For frozen models I opened and resolved https://github.com/pydantic/pydantic/issues/7784 to deal with this, so `__hash__` doesn't have to worry about it. But you definitely need to worry about it for `__eq__`, since I see no sign of deleting attributes being deprecated. And I guess maybe I need to worry about `.copy()` for `__hash__`, thanks for catching that. Though I do wonder what the point of a model with missing fields is, whether it's reached by `del` or `copy`.

I think that this should still be almost as fast in the common case:

```python
try:
    # fast path
except KeyError:
    # slow path
```

**Comment by QuentinSoubeyranAqemia** [comment]
I don't think the above code is valid: deleting the attribute violates the Liskov Substitution Principle, since the class used a type annotation to promise an attribute named `a` will always be present on instances.

This is not related to pydantic: you can remove pydantic altogether from your example, and it would still violate the LSP. If I where to write the obvious, manual `__eq__` function for this class (which is essentially what `dataclasses.dataclass` does), it would also raise an `AttributeError`.

Though since legacy pydantic interface can produce those invalid objects, I think it is better to not hard-crash and handle those. I'll use your trick (again!) from #7786 to have a low-cost fallback.

**Comment by QuentinSoubeyranAqemia** [comment]
please review

Not sure if the benchmark should be kept in the code-base. Also not sure where, or if, to add a mention of this in the docs.

**Comment by alexmojaki** [comment]
I think you must have a different version of `black` installed, or it's pointing to the wrong configuration, causing a bunch of code to get reformatted, maybe to fit in a shorter line length. Then CI doesn't complain about it because of the [new trailing commas](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#the-magic-trailing-comma). But you should probably revert the formatting-only changes.

**Comment by QuentinSoubeyranAqemia** [comment]
Thanks for catching that! That was the black VScode extension doing, it has its own configuration

**Comment by alexmojaki** [comment]
Also need to revert changes in tests/test_edge_cases.py

**Comment by QuentinSoubeyranAqemia** [comment]
Thanks again, I forgot about that file

**Comment by QuentinSoubeyranAqemia** [comment]
I cleaned the commit history with reset+re-adding things

File: pydantic/main.py
Comment (alexmojaki) [inline pydantic/main.py:932]: Did you look at checking `self.__dict__.keys() != other.__dict__.keys()` as suggested in https://github.com/pydantic/pydantic/issues/7444#issuecomment-1753091873?

File: pydantic/main.py
Comment (QuentinSoubeyranAqemia) [reply pydantic/main.py:930]: We could do that, but I'm not sure it helps much. I'll wait on the feedbacl from @lig

File: pydantic/main.py
Comment (QuentinSoubeyranAqemia) [reply pydantic/main.py:932]: What about the following dict:
- `self.__dict__ == {"a_field": 0, "_not_a_field": 1}`
- `other.__dict__ == {"a_field": 0, "_not_a_field": 99}`
?

They have the same keys, but we still want to trigger the detailed comparison with itemgetter.

Comment (QuentinSoubeyranAqemia) [reply pydantic/main.py:932]: It could work with a more detailed check on keys taking into account the model fields, something like
`not (self.__dict__.keys() == other.__dict__.keys() == cls.model_fields.keys())`.

I'll need to think about this, but this is a good idea IMHO

File: pydantic/main.py
Comment (QuentinSoubeyranAqemia) [reply pydantic/main.py]: An easy path could be to use a function to retrieve the getter and a `functools.lru_cache` to keep the getter alive? Given the getter just stores the passed tuple, I'm not sure it would be faster

File: pydantic/main.py
Comment (QuentinSoubeyranAqemia) [reply pydantic/main.py]: Added a fast-path for standard-case (i.e. no additional keys in __dict__ to ignore) unequal items. Previously, if the two __dict__ compared unequal, we would always trigger the slow __dict__ filtering code. This no longer happens when there are no keys to ignore.

**Comment by QuentinSoubeyranAqemia** [comment]
Added a fast-path for standard-case (i.e. no additional keys in `__dict__` to ignore) unequal items. Previously, if the two `__dict__` compared unequal, we would always trigger the slow `__dict__` filtering code. This no longer happens when there are no keys to ignore.

**Comment by QuentinSoubeyranAqemia** [comment]
Good to know! I've seen some people doing review commit-by-commit, so I try to have commits represent steps that would help the reviewer if they do it that way

File: pydantic/main.py
Comment (alexmojaki) [inline pydantic/main.py:927]: Did you check if this actually improves performance? It was just a suggestion, and that was before you pointed out that my idea was too simple.

Comment (alexmojaki) [reply pydantic/main.py:927]: Actually there's no need to check, having the fast path is obviously a lot better when the values are big and slow to compare.

File: pydantic/main.py
Comment (alexmojaki) [reply pydantic/main.py]: Now that `model_fields` is also being used for the latest fast path, I think it's safe to leave it.

File: pydantic/main.py
Comment (alexmojaki) [inline pydantic/main.py:73]:
```suggestion
_SENTINEL = object()
```

I think it's best to keep it simple.

File: pydantic/main.py
Comment (alexmojaki) [inline pydantic/main.py:76]: We should think about either combining our PRs now or have an idea of how we might clean up duplicate code after they both merge. This class is essentially duplicated in https://github.com/pydantic/pydantic/pull/7786/files#diff-ea75a1ef1da6bb38913477e979605f6a05e2ffe8c28fe47d9b798c89ba436c7cR422-R427 so we should agree on what it should look like. IMHO the type annotations are overkill for something so simple. I also think the `__contains__` and `__repr__` should probably be removed. `__contains__` in particular just seems to be a quirk of the type checker and gives the reader the false impression that it's actually being used. If you think they should be kept, add `# pragma: no cover` to avoid reducing test coverage.

File: pydantic/main.py
Comment (QuentinSoubeyranAqemia) [reply pydantic/main.py:927]: Yup ! And the benchmark shows that the `itemgetter()` route does induce an overhead when it is triggered (left column), so better avoid it when possible

**Comment by QuentinSoubeyranAqemia** [comment]
> We should think about either combining our PRs now or have an idea of how we might clean up duplicate code after they both merge

Fully agree. Two ways we can do this IMHO:
1. We merge our PR into a single one
2. We agree on a given implementation & naming of the helper, put it in `_internals._utils` at the same place, and whoever gets merged second resolve the conflicts

Do you have a preference @alexmojaki ?

In both case we'll share an implementation, so we need to agree on it. Current points of disagreement to resolve:
- The classes aren't quite duplicated as of now, because we have slightly different requirements for the default value:
  - you need a hashable object, so None is fine for you
  - I need a value unequal to anything the user might access, so `None` doesn't work. `object()` works well and is the canonical sentinel, but is not hashable (and is not typable, hence why I used the trick found on a MyPy issue).
  
  IIRC an enum is hashable so the enum version cover both grounds. Orwe can ask for the default in `__init__`. This probably depends on the second point, whether to keep full typing with enum or not (see below).
- Naming. I changed the name because I found the `Dict` name misleading: the class is not a `dict` and doesn't implement the full interface either. No strong feeling here
- `__contains__` is indeed a quirk of how `operator.itemgetter` is defined in [typeshed](https://github.com/python/typeshed/blob/2f8d01e9ecc6eaf9f7eb1f9bc96682070f1edfe1/stdlib/_operator.pyi#L129-L131) using [`SupportsGetItem`](https://github.com/python/typeshed/blob/2f8d01e9ecc6eaf9f7eb1f9bc96682070f1edfe1/stdlib/_typeshed/__init__.pyi#L133-L136). I'm not sure why `SupportsGetItem` also requires `__contains__`. After some spelunking, I found https://github.com/python/mypy/issues/13713 (linked from https://github.com/python/typeshed/pull/8785). Not sure I understand what is going on there.
I suppose we could hide it away in a `if TYPE_CHECKING` block like is done at over places in `pydantic/main.py`?
- `__repr__` I wrote for exceptions (originally I used a dataclass for the proxy to save implementing `__init__`), to get a clearer idea of what is going on. Having a stacktrace mention a `<_SafeGetItemProxy at 0x.....>` is probably more obscure than seeing the wrapped `__dict__`. No strong feeling either. And we could use a dataclass again to have it for free along with the trivial `__init__`

> I think it's best to keep it simple.

Indeed. If that is used in a single location, I agree that the typing is too much and should be opted-out with `Any`.

If we have a shared implementation in `_internals._utils` (be it by merging PRs or conflict resolution), I guess whether we want complete typing or not is a call for the pydantic dev & maintainers. They'll have the burden of supporting the code and its usage, and I would follow what they prefer having to maintain.

Since you wrote bare-bones and I wrote full-typing, I guess they can compare and tell us what they prefer?

**Comment by alexmojaki** [comment]
No preferences, happy to wait for reviewers, just note that `object()` is hashable.

**Comment by alexmojaki** [comment]
@QuentinSoubeyranAqemia I was about to ping the reviewers for both our PRs but I just noticed you have some merge conflicts, please can you resolve those?

**Comment by QuentinSoubeyranAqemia** [comment]
Will do asap, when I get a moment

File: pydantic/main.py
Comment (dmontagu) [inline pydantic/main.py:924]:
```suggestion
            # We don't want to trigger unnecessary costly filtering of __dict__ on all unequal objects, so we return
```

File: pydantic/main.py
Comment (dmontagu) [inline pydantic/main.py:916]:
```suggestion
            # We'll perform a fast check first, and fallback only when need
```

File: pydantic/main.py
Comment (dmontagu) [inline pydantic/main.py:931]:
```suggestion
            # Resort to costly filtering of the __dict__ objects
```

File: pydantic/main.py
Comment (dmontagu) [inline pydantic/main.py:933]:
```suggestion
            # NOTE: Contrary to standard python class and instances, when the Model class has default value for an
```

File: pydantic/main.py
Comment (dmontagu) [inline pydantic/main.py:934]:
```suggestion
            # attribute and the model instance doesn't have a corresponding attribute, accessing the missing attribute
```

**Comment by dmontagu** [comment]
I'm prepared to approve but — @alexmojaki, as you noted in a comment above — it might be good to merge https://github.com/pydantic/pydantic/pull/7786 first and then rework this PR as appropriate to reduce code duplication.

**Comment by sydney-runkle** [comment]
Agreed with @dmontagu, let's rework after we merge #7786.

See my other comment here: https://github.com/pydantic/pydantic/pull/7786#issuecomment-1806334110

**Comment by QuentinSoubeyranAqemia** [comment]
Thanks for the reviews and plan!

A question that was left pending: do you want to have:
- fully-typed code, with the enum trick and complete typing of the dict proxy, so that usage is checked by MyPy
- OR; minimal code, where we use `type: ignore` and `Any` to skip typing, to reduce code footprint
?

Both have trade-offs in term of maintenance burden, and I'd prefer to rework the code in the flavour you prefer once #7786 is merged

**Comment by sydney-runkle** [comment]
@alexmojaki,

I've merged #7786, so feel free to continue work here 👍. Thanks!

**Comment by QuentinSoubeyranAqemia** [comment]
Thanks, I'll adapt the PR to the merging of #7786 when I get a moment !

**Comment by QuentinSoubeyranAqemia** [comment]
@alexmojaki @dmontagu @sydney-runkle 
This is ready for review
I have adapted to #7786 merge, and shared the code.

I went with fully-typed flavor, so that correct usage of the `SafeGetItemProxy` utility can be checked by MyPy. If you'd prefer a non-typed with `# type: ignore` comments, or something in-between, feel free to ask in the review and I'll adapt !
I'm really not sure which is easier to maintain

**Comment by dmontagu** [comment]
@QuentinSoubeyranAqemia If I'm understanding correctly, it looks like we only use the `SafeGetItemProxy` to do comparisons and equality checks, and not to actually retrieve data in APIs where knowing the returned value's type is important. Given that, I think we should drop the use of the enum and generic, and just annotate `SafeGetItemProxy.__getitem__` as returning `Any`. In a future where we try to use `SafeGetItemProxy` to retrieve data we can change it to be more type safe, but as is, I think the complexity isn't worth it given I don't think we are using it anywhere that the values aren't of type `Any` anyway, and I think it adds a lot of cognitive overhead to the implementation.

File: pydantic/_internal/_utils.py
Comment (dmontagu) [inline pydantic/_internal/_utils.py:377]: I would suggest changing this to something like:
```python
import dataclasses
import typing
from typing import Any

from ._internal import _decorators, _internal_dataclass


_SENTINEL = object()

@dataclasses.dataclass(**_internal_dataclass.slots_true, frozen=True)
class SafeGetItemProxy:
    """Wrapper redirecting `__getitem__` to `get` with a sentinel value as default
    This makes is safe to use in `operator.itemgetter` when some keys may be missing
    """

    wrapped: dict[str, Any]

    def __getitem__(self, __key: str) -> Any:
        return self.wrapped.get(__key, _SENTINEL)

    # required to pass the object to operator.itemgetter() instances due to a quirk of typeshed
    # https://github.com/python/mypy/issues/13713
    # https://github.com/python/typeshed/pull/8785
    # Since this is typing-only, hide it in a typing.TYPE_CHECKING block
    if typing.TYPE_CHECKING:

        def __contains__(self, __key: str) -> bool:
            return self.wrapped.__contains__(__key)
```
Obviously you won't need all those imports etc.

Comment (QuentinSoubeyranAqemia) [reply pydantic/_internal/_utils.py:377]: That makes perfect sense, good point that this is only used at place where types are already Any. I'll update.

Should I merge the `_EMPTY` and `_SENTINEL` that are in that file? Looks like I can just reuse `_EMPTY`, perhaps renaming it?

Comment (QuentinSoubeyranAqemia) [reply pydantic/_internal/_utils.py:377]: What do you think about defining `__slots__` manually instead of using the special switch? IIRC it would work for all versions of python and provide the slightly faster attribute access?

**Review by dmontagu — APPROVED** [review]

**Review by sydney-runkle — APPROVED** [review]
@QuentinSoubeyranAqemia,

Thanks again. Following up on your typing question - the "fully-typed flavor" is great. This generally makes for cleaner code, and imo makes things easier to maintain as types are more clear.

**Comment by brasie** [comment]
Hi, thanks for this fix.  Do you have an estimate when this will make it into a release?  I was surprised it didn't make it into v2.5.2.

**Comment by sydney-runkle** [comment]
@brasie,

This will be included in `v2.6.0` - we only included things in the `2.5` patch releases that fixed bugs introduced by `2.5`!

**Comment by brasie** [comment]
I see, thanks for the update.

**Pull request merged**
